### PR TITLE
Extend pytorch multi tensor rmsprop optimizer to support  lr_in_momentum and decoupled_decay.

### PR DESCRIPTION
**Pull request by lly-zero-one**
Summary: Switch to use the multi-tensor version RMSProp optimizer in the classy vision flow and also make the numeric match with old one

Test Plan: Flow canary: f223946625

Differential Revision: D24102016

### CODE REVIEW

File: torch/optim/_multi_tensor/rmsprop.py [hunk 1]
```diff
@@ -28,10 +28,12 @@ class RMSprop(Optimizer):
         centered (bool, optional) : if ``True``, compute the centered RMSProp,
             the gradient is normalized by an estimation of its variance
         weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        lr_in_momentum (bool, optional): learning rate scaling is included in the
+            momentum buffer update
 
     """
 
-    def __init__(self, params, lr=1e-2, alpha=0.99, eps=1e-8, weight_decay=0, momentum=0, centered=False):
+    def __init__(self, params, lr=1e-2, alpha=0.99, eps=1e-8, weight_decay=0, momentum=0, centered=False, lr_in_momentum=False):
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/rmsprop.py:36]: This is not standard in pytorch, and doesn't seem like a standard type of argument for optimizers. What are alternatives?

Is the goal only to improve performance? In that case it would be good to have a benchmark.

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -384,7 +384,7 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
       // Check for errors and throw appropriate exception.
       checkAndThrowException();
       std::this_thread::sleep_for(
-          std::chrono::milliseconds(kSynchronizeBusyWaitMillis));
+          std::chrono::microseconds(kSynchronizeBusyWaitMicros));
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:387]: We hope to get rid of this busy wait in https://github.com/pytorch/pytorch/pull/45236, but I guess its fine to land this change if it is something urgent for classy vision.

Comment (lly-zero-one) [reply torch/lib/c10d/ProcessGroupNCCL.cpp:387]: Sorry, I was not aware this file is pulled into this PR. I will remove it.

File: torch/optim/_multi_tensor/rmsprop.py (hunk 1 above)
Comment (lly-zero-one) [reply torch/optim/_multi_tensor/rmsprop.py:36]: Thanks, make sense. I will post the perf result.

**Comment by dzhulgakov** [comment]
Can you update the PR description and maybe include a standalone benchmark results? (torch.utils.benchmark might be helpful)

File: torch/optim/_multi_tensor/rmsprop.py (hunk 1 above)
Comment (lly-zero-one) [reply torch/optim/_multi_tensor/rmsprop.py]: @vincentqb  could you help to review the extension and also the numeric change?

File: torch/lib/c10d/ProcessGroupNCCL.cpp (hunk 2 above)
Comment (pritamdamania87) [reply torch/lib/c10d/ProcessGroupNCCL.cpp]: No worries, although I was wondering if we were planning on making these changes as part of a separate PR?

**Comment by vincentqb** [comment]
> Summary: Switch to use the multi-tensor version RMSProp optimizer in the classy vision flow and also make the numeric match with old one

@lly-zero-one -- I would like more clarity about the switch to multi-tensor, and "make the numeric match with old one". Are you saying the implementation of multi-tensor doesn't match the implementation without? If this is so, then we need to add a test that can confirm that though it is hard to do in the open source without the internal models. We can sync up offline for this.

File: torch/optim/_multi_tensor/rmsprop.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/_multi_tensor/rmsprop.py]: For the API change, I do not see a reason to add this in the API of RMSProp. No other optimizers have that.

If there is a runtime speed improvement for running the step, I don't believe this would justify adding this API, since this is still a change in the RMSProp algorithm we have.

As I mention in [comment](https://github.com/pytorch/pytorch/pull/46118#issuecomment-718264930), if there is a discrepancy between RMSProp with and without multitensor, this needs to be investigated.

Comment (vincentqb) [reply torch/optim/_multi_tensor/rmsprop.py]: The change relates to #23796, and the difference between pytorch and tensorflow -- and not between the original implementation and the multi-tensor.

**Comment by ngimel** [comment]
@vincentqb it looks like there's sufficient demand for tf-style RMSProp optimizer and we are doing pytorch users a disservice by not offering it. I understand that for bc reasons we may be reluctant to change default RMSProp behavior, but then it makes sense to have RMSProp_tf optimizer (name can be anything) that does what's requested in #23796 and what @rwightman's linked optimizer does.

**Comment by lly-zero-one** [comment]
I could add an option to support the two versions.

**Comment by lly-zero-one** [comment]
> @vincentqb it looks like there's sufficient demand for tf-style RMSProp optimizer and we are doing pytorch users a disservice by not offering it. I understand that for bc reasons we may be reluctant to change default RMSProp behavior, but then it makes sense to have RMSProp_tf optimizer (name can be anything) that does what's requested in #23796 and what @rwightman's linked optimizer does.

Some comments from internal team: 
```
Seems that the pytorch rms requires a much smaller learning rate.  Currently the typical lr for tf_rms is ~0.1-0.2, while using such learning rate in pytorch_rms would lead to divergence.  I don't have extensive experiments on pytorch_rms.

We have reproduced many sota models with tf_rms.  But for pytorch_rms, some efforts are required to tune the hyper-parameters.
```

**Comment by ngimel** [comment]
@vincentqb what's required to move this PR forward and provide rms version consistent with tf behavior? It seems to be requested internally and externally.

**Comment by vincentqb** [comment]
Thanks for the pull request @lly-zero-one. As far as I can see, there are mainly three changes requested in this PR.

- `eps_inside_sqrt: see #23796. As mentioned [above](https://github.com/pytorch/pytorch/pull/46118#issuecomment-718275842) by @nikitaved I would be ok with adding this parameter to RMSProp. We can wait to see if demand arises for other optimizer, e.g. Adam.
- Decoupled weight decay: already implemented as part of AdamW (see also [paper](https://arxiv.org/pdf/1711.05101.pdf)), and should not be part of this PR.
- `lr_in_momentum` which should also be in a separate PR.

@lly-zero-one, I will close this pull request for now, but please feel free to open an issue for those points that are still needed.

File: torch/optim/_multi_tensor/rmsprop.py [hunk 3]
```diff
@@ -78,7 +102,7 @@ def step(self, closure=None):
                         raise RuntimeError('RMSprop does not support sparse gradients')
 
                     grads.append(p.grad)
-                    params_with_grad.append(p)
+                    params_with_grad.append(p.data)
```
Comment (vincentqb) [inline torch/optim/_multi_tensor/rmsprop.py:105]: We recently cleaned up the use of `.data` inside optimizers. Are you aware of its interaction with multitensor?

**Comment by rwightman** [comment]
@vincentqb decoupled decay works perfectly well with rmsprop and some other optimizers besides the AdamW/SGDW of that paper. Since I was just working on some JAX code fixing up the Flax RMSProp, it's common to see optimizers there where any weight_decay applied with the optimizer is (usually) decoupled decay and L2 penalty (equiv to weight_decay here) is applied outside of the optimizer.

Since it doesn't look like these changes are going anywhere fast, I think I'll tackle the multi-tensor variant in `timm` ... been using it a lot so probably worthwhile. It should be noted that this changeset was still short one important difference for reproducing Google papers using RMSProp + TF, rms state init.

**Comment by arikanev** [comment]
Are there any plans to revive this or do we just have to use for example, @rwightman methodology?
